packages/apollo-codegen-typescript/src/codeGeneration.ts

```diff
@@ -246,6 +335,68 @@ export class TypescriptAPIGenerator extends TypescriptGenerator {
     this.scopeStackPop();
   }
 
+  public getGlobalTypesUsedForOperation(doc: Operation | Fragment, context: CompilerContext) {
+    let docTypesUsed: GraphQLType[] = [];
+
+    if (doc.hasOwnProperty('operationName')) {
+      const operation = doc as Operation;
+      docTypesUsed = operation.variables.map(({ type }) => type);
+    }
+
+    const reduceTypesForDocument = (
+      nested: SelectionSet,
+      acc: GraphQLType[]
+    ) => {
+      acc = nested.selections
+        .reduce((selectionAcc, selection) => {
+          switch (selection.kind) {
+            case 'Field':
+            case 'TypeCondition':
+              if (selection.selectionSet) {
+                selectionAcc = reduceTypesForDocument(selection.selectionSet, selectionAcc);
```

Comment:
Updating `selectionAcc` and returning later introduces a bit of unnecessary mutability. Could you just directly return the additional types here and return `selectionAcc` for the default case?